hubble: implement peer service, enable it locally

rolinh opened the pull request:
Please, see individual commits for details.

rolinh committed:
hubble: add peer module that implements the peer gRPC server interface

This commit adds a new peer module for hubble that implements the peer
gRPC server interface.

This new gRPC service notifies of nodes being added, deleted or
modified. This information is obtained by subscribing to the cilium node
manager via a new datapath.NodeHandler interface implementation
(peer.handler).

This new service will be consumed by hubble proxy in order to connect to
hubble peers. It is intended to only be served over a local unix domain
socket.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/server: fix default options

This commits removes the global variable DefaultOptions which contained
no defaults but was just initializing the listeners map.
The map initialization is done in the constructor instead in order to
avoid problems with shallow copy of default options (a 2nd call to
NewServer would use modified defaults).

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
hubble/server: use interface for service options

Make sure to only require the grpc interface definition for service
options. The server is a simple grpc server which serves 1 or more
services and it doesn't need to know about the internal of each service
implementation. This change makes this explicit.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

rolinh committed:
daemon: enable hubble peer gRPC service on local unix socket

This commits enables the peer service for the hubble gRPC server that
listens on the local unix domain socket. This will allow node discovery
for the peer proxy.

For all other hubble listen addresses, the peer service is not enabled.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

michi-covalent commented on pkg/hubble/peer/service.go:
```diff
@@ -0,0 +1,73 @@
+				return nil
+			case cn := <-h.C:
+				if err := stream.Send(cn); err != nil {
+					//FIXME: retry?
```
maybe start with logging a warning

gandro replied:
Logging is probably sufficient for now. Also, consider checking for `io.EOF`. It means that the client has disconnected. 

However, I'm not sure if we expect clients to connect and disconnect from `Notify`often. I guess in this particular case, a client disconnecting is something which shouldn't happen except during shutdown? So then I guess the code  it's fine as.

rolinh replied:
> However, I'm not sure if we expect clients to connect and disconnect from Notify often. I guess in this particular case, a client disconnecting is something which shouldn't happen except during shutdown? So then I guess the code it's fine as.

I think you're right, it's fine as is.

michi-covalent commented on pkg/hubble/peer/peer.go:
```diff
@@ -0,0 +1,95 @@
+}
+
+// FromChangeNotification creates a new Peer from a ChangeNotification.
+func FromChangeNotification(cn *peerpb.ChangeNotification) *Peer {
```
who calls this function?

rolinh replied:
It is used by ~~hubble-proxy~~ hubble-relay. Well, only locally on my dev machine for now but in a follow-up PR for hubble-relay.

michi-covalent reviewed (commented):
the change looks good, but it wasn't clear to me how i can test this in my cluster. i guess nobody is actually calling `rpc Notify(NotifyRequest)` yet?

aanm commented on pkg/hubble/peer/peer.go:
```diff
@@ -0,0 +1,95 @@
+						Port: p,
+					}
+				} else {
+					addr = &net.TCPAddr{
+						IP:   ip,
+						Port: defaults.ServerPort,
+					}
```
This is a no-op because we are overwriting `addr` on line 84.

rolinh replied:
Good catch. It wasn't covered by the unit tests either. This is now fixed.

aanm commented on pkg/hubble/peer/service.go:
```diff
@@ -0,0 +1,73 @@
+// TODO: implement tests. This requires turning manager.Manager into an
+// interface so that it can be mocked for the purpose of testing.
+type Service struct {
+	stop chan (struct{})
```
These parentheses are redundant

aanm commented on pkg/hubble/peer/service.go:
```diff
@@ -0,0 +1,73 @@
+// interface so that it can be mocked for the purpose of testing.
+type Service struct {
+	stop chan (struct{})
+	mgr  *manager.Manager
```
Nit, consider using an interface instead of implementation.

rolinh replied:
Yup, this is required for testing, like mentioned in the struct comment just above. If it's fine with you, I'll add an interface in `pkg/node/manager/manager.go`.

aanm replied:
The interface could be defined here, not in `pkg/node/manager/manager.go`, this allows you to not need to import `pkg/node/manager` in this package.

rolinh replied:
I've added the interface with this commit faae24149afbb9da49688945123c09a79716f903 already and implemented tests of the service.
I think that importing the interface from `pkg/node/manager` is not so bad because it's otherwise really hard to understand who implements it if you're not familiar enough with cilium's codebase.

aanm replied:
@Rolinh there are tools built for that purpose [0] so there's no point in keeping the interfaces in the same place as 1 of the implementations. What happens if in the future we will have a different node manager that implements the same interface? Where will the interface live?

[0] https://docs.google.com/document/d/1_Y9xCEMj5S-7rv2ooHpZNH15JgRT5iM742gJkw5LtmQ/edit#heading=h.f6hyappclfor

rolinh replied:
I don't mind moving the interface to `pkg/hubble/peer` if you have strong feelings about it; I don't think it matters much.

rolinh replied:
> The interface could be defined here, not in pkg/node/manager/manager.go, this allows you to not need to import pkg/node/manager in this package.

BTW, by moving the interface to `pkg/hubble/peer` I could save on importing `pkg/node/manager` but then I'll have to import `pkg/datapath` as it's required for the interface definition. I'm not "saving" on import anyway.

gandro commented on pkg/hubble/server/serveroption/option.go:
```diff
@@ -34,8 +35,8 @@ import (
 type Options struct {
 	Listeners       map[string]net.Listener
 	HealthService   *health.Server
```
If you also want to use an interface here, I believe this one would be it: https://godoc.org/google.golang.org/grpc/health/grpc_health_v1#HealthServer

rolinh replied:
Oh yes, good point, thanks!

gandro commented on pkg/hubble/defaults/defaults.go:
```diff
@@ -0,0 +1,24 @@
+const (
+	// ServerPort is the default port for hubble server when a provided
+	// listen address does not include one.
+	ServerPort = 4244
```
It doesn't have to be done now (since we are still developing that feature and don't want people use it in production), but once we add documentation, we should not forget to update the list here: https://docs.cilium.io/en/v1.7/install/system_requirements/#firewall-rules

michi-covalent replied:
does it work with non-default port at all? how does it know which port to connect to if the notification doesn't include the port? 

```
{
  "name": "runtime1",
  "address": "10.0.2.15",
  "type": "PEER_ADDED"
}
```

rolinh replied:
> does it work with non-default port at all? how does it know which port to connect to if the notification doesn't include the port?

That's a valid point. At the moment, it doesn't work with non-default ports. Not sure how we could change this because such information is not available to the node manager.

Eventually, this also means that the `--hubble-listen-addresses` doesn't make a ton of sense anymore and should probably be changed to `--enable-hubble-remote-service` or something like that.

michi-covalent replied:
ok i guess we'll figure something out later.

gandro commented on pkg/hubble/peer/service.go:
```diff
@@ -0,0 +1,73 @@
+			case <-s.stop:
+				return nil
+			case cn := <-h.C:
+				if err := stream.Send(cn); err != nil {
```
`stream.Send()` can block (it does I/O). Which means the unbuffered `h.C` channel causes the NodeManager to be blocked as well in `h.Node{Add,Update,Delete}`.

aanm replied:
Why isn't a `context.Context` part of the `Send` parameters?

rolinh replied:
Because the [`ServerStream` interface](https://godoc.org/google.golang.org/grpc#ServerStream) doesn't use a `context.Context`. `SendMsg()` is used by `Send()`, as can be seen in the generated code in `peer.pb.go`:
```go
type Peer_NotifyServer interface {
	Send(*ChangeNotification) error
	grpc.ServerStream
}

type peerNotifyServer struct {
	grpc.ServerStream
}

func (x *peerNotifyServer) Send(m *ChangeNotification) error {
	return x.ServerStream.SendMsg(m)
}
```
The interface defines a context but it's global to `ServerStream`.

rolinh replied:
@gandro I have added a test that simulates a blocked send (`TestService_NotifyWithBlockedSend`) and it failed (which is good as it tests what we want).

To fix the problem and have the test pass, I've added small buffer (of size 10 by default but this is configurable through service options) that, when full, triggers the server to stop steaming to the client.
This eventually pushes the problem back to the client which will have to call `Notify()` again in such a case. I don't think it's a problem overall because upon calling `Notify()`, the client receives the list of all peers before receiving continuous updates so maintaining a local cache client-side is easy.

Please, have a look again at the `Notify()` implementation and tell me if you think this way of handling the problem is OK.

gandro replied:
Awesome, yes, the code itself looks good! :tada: We have to get a bit of experience with this, because I honestly have no clue how big the buffer size should be.

rolinh replied:
> We have to get a bit of experience with this, because I honestly have no clue how big the buffer size should be.

Neither do I. I assumed that since the connection is over a local unix domain socket, a buffer of 10 should be sufficient. I also don't want to make it unnecessarily large because if the caller can't keep up then there's probably a problem elsewhere.

rolinh committed:
node/manager: add Notifier interface

This interface will make it easier to implement tests, notably for the
peer service that uses the node manager.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

gandro reviewed (changes requested):
Awesome! I do see an issue with blocking the Node Manager which I think must be addressed, but all my other comments are just comments and don't require any action.

rolinh committed:
hubble: move FakeGRPCServerStream to testutils package

This is useful for any package that needs to implement gRPC tests, thus
move it to a more generic location.

Signed-off-by: Robin Hahling <robin.hahling@gw-computing.net>

aanm commented on pkg/hubble/peer/handler.go:
```diff
@@ -0,0 +1,117 @@
+// Once not used anymore, Close must be called to free resources.
+type handler struct {
+	stop chan struct{}
+	C    chan (*peerpb.ChangeNotification)
```
redundant parentheses

christarazi reviewed (approved):
LGTM!
